File: Go/Ratelimiter-go/leakybucket.go
package main

import (
	"sync"
	"time"
)

const (
	leakyRateLimit    = 5
	leakyLeakRateInMs = 1000
)

type LeakyBucket struct {
	Limit    int
	LeakRate int64
	Requests int
	LastLeak int64
}

func NewLeakyBucket(limit int, leakRate int64) *LeakyBucket {
	return &LeakyBucket{Limit: limit, LeakRate: leakRate}
}

func (b *LeakyBucket) leak() {
	now := time.Now().UnixMilli()
	timeElapsed := now - b.LastLeak
	leakedRequests := int(timeElapsed / b.LeakRate)
	b.Requests = max(0, b.Requests-leakedRequests)
	b.LastLeak = now
}

func (b *LeakyBucket) AllowRequest() bool {
	b.leak()
	if b.Requests < b.Limit {
		b.Requests++
		return true
	}
	return false
}

type LeakyBucketLimiter struct {
	mu      sync.Mutex
	buckets map[string]*LeakyBucket
}

func NewLeakyBucketLimiter() *LeakyBucketLimiter {
	return &LeakyBucketLimiter{buckets: make(map[string]*LeakyBucket)}
}

func (l *LeakyBucketLimiter) AllowRequest(userIP string) bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	if _, ok := l.buckets[userIP]; !ok {
		l.buckets[userIP] = NewLeakyBucket(leakyRateLimit, leakyLeakRateInMs)
	}
	return l.buckets[userIP].AllowRequest()
}

File: JavaScript/Ratelimiter/leakyBucket.js
const RATE_LIMIT = 5;
const LEAK_RATE_IN_MS = 1000; // Leak 1 request per second
const userBuckets = new Map();

class LeakyBucket {
  constructor(limit, leakRate) {
    this.limit = limit;
    this.leakRate = leakRate;
    this.requests = 0;
    this.lastLeak = Date.now();
  }

  leak() {
    const now = Date.now();
    const timeElapsed = now - this.lastLeak;

    const leakedRequests = Math.floor(timeElapsed / this.leakRate);

    this.requests = Math.max(0, this.requests - leakedRequests);
    this.lastLeak = now;
  }

  allowRequest() {
    this.leak();
    if (this.requests < this.limit) {
      this.requests++;
      return true;
    }
    return false;
  }
}

// Leaky Bucket Middleware
const leakyBucketRateLimiter = (req, res, next) => {
  const userIP = req.ip;

  if (!userBuckets.has(userIP)) {
    userBuckets.set(userIP, new LeakyBucket(RATE_LIMIT, LEAK_RATE_IN_MS));
  }

  const bucket = userBuckets.get(userIP);

  if (bucket.allowRequest()) {
    next();
  } else {
    res
      .status(429)
      .json({ message: 'Too many requests. Please try again later.' });
  }
};

module.exports = leakyBucketRateLimiter;
